programs/cleos/main.cpp

```diff
@@ -1153,6 +1156,263 @@ int main( int argc, char** argv ) {
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Fail to parse transaction JSON '${data}'", ("data",proposed_transaction))
+      transaction proposed_trx = trx_var.as<transaction>();
+
+      auto arg= fc::mutable_variant_object()
+         ("code", proposed_contract)
+         ("action", proposed_action)
+         ("args", trx_var);
+
+      auto result = call(json_to_bin_func, arg);
+      //std::cout << "Result: "; fc::json::to_stream(std::cout, result); std::cout << std::endl;
+
+      bytes proposed_trx_serialized = result.get_object()["binargs"].as<bytes>();
+
+      vector<permission_level> reqperm;
+      try {
+         reqperm = requested_perm_var.as<vector<permission_level>>();
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Wrong requested permissions format: '${data}'", ("data",requested_perm_var));
+
+      try {
+         transaction_perm_var.as<vector<permission_level>>();
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Wrong transaction permissions format: '${data}'", ("data",transaction_perm_var));
+
+      auto accountPermissions = get_account_permissions(tx_permission);
+      if (accountPermissions.empty()) {
+         if (!proposer.empty()) {
+            accountPermissions = vector<permission_level>{{proposer, config::active_name}};
+         } else {
+            EOS_THROW(tx_missing_auth, "Authority is not provided (either by multisig parameter <proposer> or -p)");
+         }
+      }
+      if (proposer.empty()) {
+         proposer = name(accountPermissions.at(0).actor).to_string();
+      }
+
+      arg = fc::mutable_variant_object()
+         ("code", "eosio.msig")
+         ("action", "propose")
+         ("args", fc::mutable_variant_object()
+          ("proposer", proposer )
+          ("proposal_name", proposal_name)
+          ("requested", requested_perm_var)
+          ("trx", fc::mutable_variant_object()
+           ("expiration", "2018-05-01T00:00:00")
+           ("region", 0)
+           ("ref_block_num", 0)
+           ("ref_block_prefix", 0)
+           ("max_net_usage_words", 0)
+           ("max_kcpu_usage",0)
+           ("delay_sec", 0)
+           ("context_free_actions", fc::variants() )
+           ("actions", fc::variants{ fc::mutable_variant_object()
+                                     ("account", proposed_contract)
+                                     ("name", proposed_action)
+                                     ("authorization", transaction_perm_var)
+                                     ("data", proposed_trx_serialized)
+                 }
+           )
+          )
+         );
+      result = call(json_to_bin_func, arg);
+      send_actions({chain::action{accountPermissions, "eosio.msig", "propose", result.get_object()["binargs"].as<bytes>()}});
+   });
+
+   //resolver for ABI serializer to decode actions in proposed transaction in multisig contract
+   auto resolver = [](const name& code) -> optional<contracts::abi_serializer> {
```

Comment:
In an effort to keep cleos as "thin" as possible I think we should consider moving all of these uses of abi_serializer that depend on chain state (`get_code_func`) to the chain RPC with a call that mirrors and is the opposite of `abi_json_to_bin`.  It is a round-trip to the RPC either way but, if we legitimize that endpoint then other non-cleos HTTP services can benefit.

Later as an optimization if we want to introduce a client side cache for the ABI we can do so but, then the ABI's would come locally.